microsoft/surface/README.md

```diff
@@ -0,0 +1,75 @@
+# Derivatives for Microsoft Surface notebooks
+
+These derivatives use the patches from the [linux-surface repo](https://github.com/linux-surface/linux-surface/tree/master/patches).
+
+## Kernel
+
+The kernel needs several patches to make it work correctly with some of the hardware on various
+Surface models, e.g. keyboard/trackpad, camera, wifi.
+
+Not all hardware is fully supported, but the
+[linux-surface feature matrix](https://github.com/linux-surface/linux-surface/wiki/Supported-Devices-and-Features#feature-matrix)
+provides details on which devices are supported on which types of machine.
+
+The kernel-specific derivations are under the `kernel/` sub-directory.
+In order to simplify maintenance of the Nix code, only the most-recent kernel patch-set is expected
+to be maintained in this repo.
+
+_*NOTE:*_Some built-in Kernel config items need to be set, that aren't set by default:
+- https://github.com/linux-surface/surface-aggregator-module/wiki/Testing-and-Installing
+
+## Firmware, Drivers and Support Tools
+
+### IPTS
+
+IPTS is used on most of the Surface range, except for Surface Go and Surface Laptop 3 (AMD version).
+
+The firmware files from the [surface-ipts-firmware repo](https://github.com/linux-surface/surface-ipts-firmware.git)
+are only needed on kernel 4.19.
+
+On kernel 5.4 or newer you may also need the user-space `iptsd` daemon (below).
+
+### WiFi
+
+For the Surface Go, please see the "Issues" sections below.
+
+### IPTSD
+
+The `iptsd` daemon works with the `intel-precise-touch` driver to convert raw touch data into events
```

Comment:
Also this should go to nixpkgs.

Reply:
The intel precise touch driver is technically split into a kernelspace part (`intel-precise-touch`) and a user-space part (`iptsd`).

The kernelspace part collects the raw capacitive touch data from the Intel Management Engine, and provides an API for the user-space to connect to.
The user-space portion does some floating-point manipulation of the raw data and provides this to the HID / input subsystem.

I believe the IPTS system is not exclusively used on MS Surface, so could be moved elsewhere in the repo, but Linux can't use the hardware without both parts working together.

Could that mean both parts should stay in this repo?

Reply:
Having it in nixpkgs allows to test for build breakages and provide a binary cache. Having it here allows to update kernel and userspace together. Which one do you prefer?

Reply:
> Having it in nixpkgs allows to test for build breakages and provide a binary cache. Having it here allows to update kernel and userspace together. Which one do you prefer?

I feel like the kernel-space and user-space code should probably stay together at this stage, because neither will work without the other;  I'm checking whether they are tightly-coupled across releases, though.
I also feel that since the `iptsd` takes a fraction of the time that a new kernel takes to build, any time saved by binary caching might not matter right now.

It seems like the scripts in `//tests/` are working pretty well, too?

In future, I assume it would be OK to deprecate and move into `nixpkgs` if that makes more sense?

Reply:
Yes. But might cause troubles with stable releases (that don't have those packages yet) - so some backport would be needed.

Reply:
> I feel like the kernel-space and user-space code should probably stay together at this stage, because neither will work without the other; I'm checking whether they are tightly-coupled across releases, though.

I've confirmation from upstream that they are expected to be released together, and that while there is some compatibility between releases, the two sides are reasonably tightly-coupled at the API.

> Yes. But might cause troubles with stable releases (that don't have those packages yet) - so some backport would be needed.

Is this referring to my "deprecate and move into `nixpkgs`" question?
If so:  I know this could be less-smooth, and (personally) think if it happens, then the kernel patches would also have to move.

So my thinking goes like this:
I expect there'll be some negotiation to get these derivations accepted in `nixpkgs` as it looks like upstream can't support every single patch-release of a kernel, anyway. 
Matching that to whichever kernel is "current" in `nixpkgs` in a way that people are comfortable with might take some negotiation.

TBH, that's one of the advantages of `nixos-hardware` : it can get some use by people (e.g. me!) while that negotiation happens.

Does that sound reasonable?

Reply:
I don't think that we want to maintain more kernel configurations in nixpkgs. We need to test kernel modules for every one of them and it becomes tedious if they break for certain kernel configurations. So this kernel might remain in nixos-hardware for a while. Maybe I can also organize merges access for you so that kernel updates can be delivered timely.